File: Tools/QuantityAndUnit.cs
using System;
using System.Globalization;
using System.Linq;

namespace VorratsUebersicht
{
    public class QuantityAndUnit
    {
        public decimal Quantity;
        public string Unit;

        public QuantityAndUnit(decimal quantity, string unit)
        {
            this.Quantity = quantity;
            this.Unit     = unit;
        }

        public static QuantityAndUnit Parse(string quantityAndUnit)
        {
            if (string.IsNullOrEmpty(quantityAndUnit))
                return null;

            QuantityAndUnit value;
            quantityAndUnit = quantityAndUnit.Trim();

            value = QuantityAndUnit.Parse(quantityAndUnit, " kg"); if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, " g");  if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, " gr"); if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, " l");  if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, " ml"); if (value != null) return value;

            value = QuantityAndUnit.Parse(quantityAndUnit, "kg"); if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, "g");  if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, "gr"); if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, "l");  if (value != null) return value;
            value = QuantityAndUnit.Parse(quantityAndUnit, "ml"); if (value != null) return value;

            return null;
        }

        public static QuantityAndUnit Parse(string quantityAndUnit, string unit)
        {
            // kg, g, l, ml
            if (!quantityAndUnit.EndsWith(unit, StringComparison.InvariantCultureIgnoreCase))
                return null;

            string quantityText = quantityAndUnit.Substring(0, quantityAndUnit.Length - unit.Length);
            unit = unit.Trim().ToLower();

            quantityText = quantityText.Replace(",", ".");
            decimal quantity;

            if (!decimal.TryParse(quantityText, NumberStyles.Any, CultureInfo.InvariantCulture, out quantity))
                return null;

            return new QuantityAndUnit(quantity, unit);
        }
    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/tools/QuantityAndUnit.kt
package de.stryi.vorratsuebersicht.tools

import java.util.Locale

class QuantityAndUnit(var quantity: Double, var unit: String) {
    //var quantity: Double = 0.00
    //var unit: String = ""

    fun quantityAndUnit(quantity: Double, unit: String) {
        this.quantity = quantity
        this.unit = unit
    }

    companion object {

        fun parse(valueText: String?) : QuantityAndUnit? {

            if (valueText.isNullOrEmpty())
                return null

            val quantityAndUnit = valueText.trim()

            var value : QuantityAndUnit?

            value = parse(quantityAndUnit, " kg"); if (value != null) return value
            value = parse(quantityAndUnit, " g");  if (value != null) return value
            value = parse(quantityAndUnit, " gr"); if (value != null) return value
            value = parse(quantityAndUnit, " l");  if (value != null) return value
            value = parse(quantityAndUnit, " ml"); if (value != null) return value

            value = parse(quantityAndUnit, "kg"); if (value != null) return value
            value = parse(quantityAndUnit, "g");  if (value != null) return value
            value = parse(quantityAndUnit, "gr"); if (value != null) return value
            value = parse(quantityAndUnit, "l");  if (value != null) return value
            value = parse(quantityAndUnit, "ml"); if (value != null) return value

            return null
        }

        fun parse(quantityAndUnit: String, unit: String): QuantityAndUnit?
        {
            // Prüfen, ob String mit der Einheit endet (case-insensitive)
            if (!quantityAndUnit.endsWith(unit, ignoreCase = true)) {
                return null
            }

            // Zahl extrahieren
            var quantityText = quantityAndUnit.substring(0, quantityAndUnit.length - unit.length)
            val normalizedUnit = unit.trim().lowercase(Locale.getDefault())

            // Komma gegen Punkt tauschen
            quantityText = quantityText.replace(",", ".")

            // Versuch, in Double umzuwandeln
            val quantity = quantityText.toDoubleOrNull() ?: return null

            return QuantityAndUnit(quantity, normalizedUnit)
        }
    }
}